pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java

```diff
@@ -0,0 +1,198 @@
+  public final static long NO_IN_FILTER_THRESHOLD = 0;
+  public final static long CARD_THRESHOLD_ONE = 1;
+
+  public final static long NO_PROCESSED_THRESH = 0;
+
+  public final static Pattern _dimensionPattern = Pattern.compile(DIMENSION_REGEX);
+  private HashSet<String> _tableNamesWithoutType;
+  private long _numEntriesScannedThreshold;
+  private long _cardinalityThreshold;
+  private long _numQueriesThreshold;
+
+  private FrequencyImpl(Builder builder) {
+    _tableNamesWithoutType = builder._tableNamesWithoutType;
+    _numEntriesScannedThreshold = builder._numEntriesScannedThreshold;
+    _cardinalityThreshold = builder._cardinalityThreshold;
+    _numQueriesThreshold = builder._numQueriesThreshold;
+  }
+
+  public static final class Builder {
+    private HashSet<String> _tableNamesWithoutType = new HashSet<>();
+    private long _numEntriesScannedThreshold = NO_IN_FILTER_THRESHOLD;
+    private long _cardinalityThreshold = CARD_THRESHOLD_ONE;
+    private long _numQueriesThreshold = NO_PROCESSED_THRESH;
+
+    public Builder() {
+    }
+
+    @Nonnull
+    public FrequencyImpl build() {
+      return new FrequencyImpl(this);
+    }
+
+    /**
+     * set the tables to work on, other tables will be filtered out
+     * @param val set of table names without type
+     * @return
+     */
+    @Nonnull
+    public Builder setTableNamesWithoutType(@Nonnull HashSet<String> val) {
+      _tableNamesWithoutType = val;
+      return this;
+    }
+
+    /**
+     * set the threshold for _numEntriesScannedInFilter, the queries with _numEntriesScannedInFilter below this will be filtered out
+     * @param val
+     * @return
+     */
+    @Nonnull
+    public Builder setNumEntriesScannedThreshold(long val) {
+      _numEntriesScannedThreshold = val;
+      return this;
+    }
+
+    /**
+     * set the cardinality threshold, column with cardinality below this will be ignored,
+     * setting a high value will force the system to ignore low card columns
+     * @param val cardinality threshold, default to 1
+     * @return
+     */
+    @Nonnull
+    public Builder setCardinalityThreshold(long val) {
+      _cardinalityThreshold = val;
+      return this;
+    }
+
+    /**
+     * set the minimum number of records scanned to give a recommendation
+     * @param val minimum number of records scanned to give a recommendation, default to 0
+     * @return
+     */
+    @Nonnull
+    public Builder setNumQueriesThreshold(long val) {
+      _numQueriesThreshold = val;
+      return this;
+    }
+  }
+
+  @Override
+  public boolean filter(AbstractQueryStats queryStats) {
+    IndexSuggestQueryStatsImpl indexSuggestQueryStatsImpl = (IndexSuggestQueryStatsImpl) queryStats;
+    long numEntriesScannedInFilter = Long.parseLong(indexSuggestQueryStatsImpl.getNumEntriesScannedInFilter());
+    return (_tableNamesWithoutType == null || _tableNamesWithoutType.isEmpty() || _tableNamesWithoutType
```

Comment:
We will not skip the numEntriesScannedThreshold check because it's (_tableNamesWithoutType == null || ... || ...) **&&** (numEntriesScannedInFilter > _numEntriesScannedThreshold)

Reply:
> why is filter method being made to check for _tableNamesWithoutType == null || _tableNamesWithoutType.isEmpty? This is going to be repeated simply for every query. Can it be done before the 4 stages begin?

Good point, I will make it in the builder of strategy.